analysis_runner/dataproc.py

```diff
@@ -46,6 +49,10 @@ def hail_dataproc_job(
 
     cluster_name = f'dataproc-{uuid.uuid4().hex}'
 
+    # Format labels
+    labels_string = re.sub('"|{|}| ', '', json.dumps(labels))
+    labels_formatted = re.sub(':', '=', labels_string)
```

Comment:
Can we build this directly from `labels` without involving JSON and REs? Are certain characters not allowed in label keys or values?

Reply:
Do you mean passing `labels` without formatting it? From what I can tell the labels need to be passed in as per the syntax in the [docs](https://cloud.google.com/dataproc/docs/guides/creating-managing-labels) i.e. 
`--labels key1=value1,key2=value2`. If we just provided the dict it looks like it will break. And yup, the characters need to be alphanumeric.

Reply:
The rules for labels are actually a bit more complicated: https://cloud.google.com/resource-manager/docs/creating-managing-labels#requirements

I'd suggest adding a link to that documentation to the parameter. We don't need to duplicate the checks in the code here, as the cluster creation will fail anyway if the labels aren't well-formed.

> Do you mean passing labels without formatting it?

No, I just meant that converting to JSON and then modifying the resulting string is a bit of a roundabout way. How about this:

```python
labels_formatted = ','.join(f'{key}={value}' for key, value in labels.items())
```
```

Reply:
> I'd suggest adding a link to that documentation to the parameter. We don't need to duplicate the checks in the code here, as the cluster creation will fail anyway if the labels aren't well-formed.

Will do! 

> No, I just meant that converting to JSON and then modifying the resulting string is a bit of a roundabout way. How about this:

Ah, I understand. That's much better than what I had, thank you!